primitives/src/witness.rs

```diff
@@ -312,6 +320,173 @@ impl<'a> Encoder for WitnessEncoder<'a> {
     }
 }
 
+/// The decoder for the [`Witness`] type.
+#[cfg(feature = "alloc")]
+pub struct WitnessDecoder {
+    /// A decoder for the initial length prefix and subsequent per-element prefixes.
+    prefix_decoder: Option<CompactSizeDecoder>,
+    /// Holds the elements.
+    buffer: Vec<Vec<u8>>,
+    /// True if the initial compact size has been read.
+    initial_length_prefix_read: bool, // I.e not the one for each element.
```

Comment:
Can we simplify the state by dropping `initial_length_prefix_read` and` element_length_prefix_read` and base on if a CompactSizeDecoder and `witness_elements` are present kinda like in VecDecoder?

Reply:
We re-use the `prefix_decoder` for both the initial and per element varints so its not a straight refactor like with the `VecDecoder`. I'm not saying there is not a better/cleaner way to write it just that its not a 2 minute refactor like the other ones were.

Reply:
Yea, hear ya, can be done later.